internal/pkg/agent/application/monitoring/process.go

```diff
@@ -82,14 +81,18 @@ func processHandler(coord *coordinator.Coordinator, statsHandler func(http.Respo
 
 		state := coord.State()
 
-		for _, c := range state.Components {
-			if matchesCloudProcessID(&c.Component, componentID) {
+		unhealthyComponent := false
+		for _, comp := range state.Components {
+			if comp.State.State == client.UnitStateFailed || comp.State.State == client.UnitStateDegraded {
+				unhealthyComponent = true
+			}
```

Comment:
Hmm, good question. @cmacknz thoughts?

Reply:
It's a public holiday today for Craig, he will reply on Tuesday.
I'm not sure restarting the full container when a component is degraded is the best thing to do but maybe I'm wrong.

Reply:
From my point of view (which could be totally wrong) we only want to fail the "liveness" check if we believe that restarting the container has a chance of fixing the problem.  For example if something in the configuration sent down from fleet caused a component to be in Degraded state, I don't want to fail the "liveness" check.  Restarting the container won't fix the issue, and we will just be constantly restarting, which makes figuring out what is wrong even harder.

So it might be worth asking the question "What situations do we have where a reboot of the container is likely to fix the problem?"

Reply:
Yes, there actually aren't many situations where a container restart will fix a problem. At the same time, we don't know the exact conditions of every possible error an input will report. Some people may also end up using this as a crude alert on agent pods needing intervention.

The nicest way to deal with this is to make it configurable. We could have several options.
1. Liveness only checks that the container can respond to an HTTP request at all. This should be the default.
2. Liveness fails when any component or unit is in the failed state.
3. Liveness fails when any component or unit is in the degraded state.
4. Liveness fails when the agent determines it is offline from Fleet (no check ins for X minutes).

We could also optionally introduce the ability to filter which components or unit names are included in these conditions, in case someone wants to limit checks to only outputs or log collection or something.

Reply:
Alright, so what we need is some config subsection in`monitoring.http` that lets us define when we actually return a `500`. 

> Liveness only checks that the container can respond to an HTTP request at all. This should be the default.

So the default should just be that the endpoint works at all, and not based on any component/unit state? That seems like it might be confusing?

Reply:
> So the default should just be that the endpoint works at all, and not based on any component/unit state? That seems like it might be confusing?

Yeah thinking a bit more we could have the default be to fail if anything is in the failed state, with the option to restrict it or loosen it to include degraded. I have a suspicion people want to solve the problem where an input can't be created at all and the container is considered "alive" even though it is sitting there not collecting the data they want silently. This makes it more useful that just hitting `/stats` by default but allows users to control what it does.

For a concrete example of how we could configure this:
```yaml
 agent.monitoring:
  http:
      # enables http endpoint
      enabled: true
      # The criteria used to determine with the /liveness endpoint should not return a 200 OK response
      # Must be one of: "failed", "degraded", "none". Defaults to "failed".
      # - "failed" returns an error when any component or unit is in the failed state. 
      # - "degraded" returns an error when any component or unit is in the degraded state.
      # - "none" always returns success as long as the agent can respond over HTTP at all.
      liveness_criteria: "failed"
```

Reply:
If you want to you can make `liveness` its own sub-object if you think it's nicer or we could add more keys in the future. Also open to a different name than `criteria` if someone can think of a better one.

```yaml
 agent.monitoring:
  http:
      # enables http endpoint
      enabled: true
      liveness: 
        # The criteria used to determine with the /liveness endpoint should not return a 200 OK response
        # Must be one of: "failed", "degraded", "none". Defaults to "failed".
        # - "failed" returns an error when any component or unit is in the failed state. 
        # - "degraded" returns an error when any component or unit is in the degraded state.
        # - "none" always returns success as long as the agent can respond over HTTP at all.
        criteria: "failed"
```

Reply:
I think if we want to provide options to adjust the result of the HTTP response it would be better if its in the HTTP request, like `/healthz?criteria=failed`.

I also had some original ideas on this in the linked issue - https://github.com/elastic/elastic-agent/issues/390#issuecomment-1400796372

Reply:
+1, query parameters are better than configuration.